### PR TITLE
[Validate] Provide endpoint for uploading custom evaluation results

### PR DESCRIPTION
**Pull request by jean-lucas**
Needs to run with the backend in the following PR: https://github.com/scaleapi/scaleapi/pull/40729

Example usage:

```python
client = nucleus.NucleusClient(API_KEY)

client.validate.create_external_eval_function('Metric_A')

# Fetch the new created functions (TODO: find a way to skip this step)
external_fns = client.validate.eval_functions.external_functions
eval_fn_a = external_fns['Metric_A']

# pick what ever you like (this slice works for me)
model_id = 'prj_bybpa3gjmjc30es761y0'
slice_id = 'slc_c9cr0hts4gvg0ehw9qbg'

# create a scenario test on the given slice with the desired evaluation functions
scenarioTest  = client.validate.create_scenario_test('some_name', slice_id, [eval_fn_a])

metric_results_A =[
    EvaluationResult(item_ref_id='PennPed00066', score=0.8, weight=0.1),
    EvaluationResult(item_ref_id='PennPed00072', score=0.7, weight=0.2),
    EvaluationResult(item_ref_id='PennPed00073', score=0.2, weight=0.3),
]

scenarioTest.upload_external_evaluation_results(eval_fn_a, metric_results_A, model_id)

```

### CODE REVIEW

File: nucleus/validate/data_transfer_objects/scenario_test_evaluations.py [hunk 1]
```diff
@@ -9,3 +11,17 @@ class EvalDetail(ImmutableModel):
 
 class GetEvalHistory(ImmutableModel):
     evaluations: List[EvalDetail]
+
+
+class EvaluationResult(ImmutableModel):
```
Comment (pfmark) [inline nucleus/validate/data_transfer_objects/scenario_test_evaluations.py:16]: Let's add the this class to the `validate/__init__.py` imports to allow for easy imports on the client side like `from nucleus.validate import EvaluationResult`

File: nucleus/validate/scenario_test.py [hunk 2]
```diff
@@ -83,9 +89,13 @@ def add_eval_function(
         Args:
             eval_function: :class:`EvalFunction`
 
+        Raises:
+            NucleusAPIError: If by adding this function, the scenario test mixes external with non-external functions
```
Comment (pfmark) [inline nucleus/validate/scenario_test.py:93]:
```suggestion
            NucleusAPIError: By adding this function, the scenario test mixes external with non-external functions which is not permitted.
```

**Review by pfmark — COMMENTED** [review]
Awesome work overall! 

Unfortunately uploading the results didn't work for me although everything else up to that step worked. I'm getting the following error in the backend:
```
[api] info: Changing user_id to default scaleapi user since internal scale user {"label":"scale-insight-api"}
[api] POST /v1/nucleus/validate/scenario_test/upload_results 404 2305.051 ms - 188
```

File: nucleus/validate/data_transfer_objects/scenario_test_evaluations.py (hunk 1 above)
Comment (jean-lucas) [reply nucleus/validate/data_transfer_objects/scenario_test_evaluations.py:16]: good point. Fixed.

**Comment by jean-lucas** [comment]
> Unfortunately uploading the results didn't work for me although everything else up to that step worked. I'm getting the following error in the backend:

Issue resolved. Need to merge the backend first. 

Thanks for taking a look.

In the backend, I will add a check on `upload_external_evaluation_results`, to fail if the eval_fn does not yet exist for the scenario test

**Comment by jean-lucas** [comment]
@pfmark when submitting results to a scenario test with an un-linked eval function, the user will see this error:
Thanks for catching this.

image

**Review by gatli — APPROVED** [review]
LGTM for now. We can change the aggregation logic later.

File: nucleus/validate/scenario_test.py [hunk 3]
```diff
@@ -174,3 +184,43 @@ def set_baseline_model(self, model_id: str):
...
+        metric_per_ref_id = {}
+        weight_per_ref_id = {}
+        aggregate_weighted_sum = 0.0
+        aggregate_weight = 0.0
+        # aggregation based on https://en.wikipedia.org/wiki/Weighted_arithmetic_mean
+        for r in results:
+            metric_per_ref_id[r.item_ref_id] = r.score
+            weight_per_ref_id[r.item_ref_id] = r.weight
+            aggregate_weighted_sum += r.score * r.weight
+            aggregate_weight += r.weight
```
[the remark this answers is not in the document]
Comment (jean-lucas) [reply nucleus/validate/scenario_test.py:211]: You're right, I missed that point.
I've made a ticket to address this: https://app.shortcut.com/scaleai/story/475163/perform-aggregation-of-results-in-the-backend

hopefully tackle it by early next week.

**Pull request merged**